iot/api-client/end_to_end_example/cloudiot_pubsub_example_mqtt_device.py

```diff
@@ -0,0 +1,242 @@
+    if not payload:
+      return
+
+    # The config is passed in the payload of the message. In this example, the
+    # server sends a serialized JSON string.
+    data = json.loads(payload)
+    if data['fan_on'] != self.fan_on:
+      # If we're changing the state of the fan, print a message and update our
+      # internal state.
+      self.fan_on = data['fan_on']
+      if self.fan_on:
+        print 'Fan turned on.'
+      else:
+        print 'Fan turned off.'
+
+
+def parse_command_line_args():
+  """Parse command line arguments."""
+  parser = argparse.ArgumentParser(
+      description='Example Google Cloud IoT MQTT device connection code.')
+  parser.add_argument(
+      '--project_id', required=True, help='GCP cloud project name')
+  parser.add_argument(
+      '--registry_id', required=True, help='Cloud IoT registry id')
+  parser.add_argument('--device_id', required=True, help='Cloud IoT device id')
+  parser.add_argument(
+      '--private_key_file', required=True, help='Path to private key file.')
+  parser.add_argument(
+      '--algorithm',
+      choices=('RS256', 'ES256'),
+      required=True,
+      help='Which encryption algorithm to use to generate the JWT.')
+  parser.add_argument(
+      '--cloud_region', default='us-central1', help='GCP cloud region')
+  parser.add_argument(
+      '--ca_certs',
+      default='roots.pem',
+      help='CA root certificate. Get from https://pki.google.com/roots.pem')
+  parser.add_argument(
+      '--num_messages',
+      type=int,
+      default=100,
+      help='Number of messages to publish.')
+  parser.add_argument(
+      '--mqtt_bridge_hostname',
+      default='mqtt.googleapis.com',
+      help='MQTT bridge hostname.')
+  parser.add_argument(
+      '--mqtt_bridge_port', default=8883, help='MQTT bridge port.')
+
+  return parser.parse_args()
+
+
+def main():
+  args = parse_command_line_args()
+
+  # Create our MQTT client and connect to Cloud IoT.
+  client = mqtt.Client(
+      client_id='projects/{}/locations/{}/registries/{}/devices/{}'.format(
+          args.project_id, args.cloud_region, args.registry_id, args.device_id))
+  client.username_pw_set(
+      username='unused',
+      password=create_jwt(args.project_id, args.private_key_file,
+                          args.algorithm))
+  client.tls_set(ca_certs=args.ca_certs)
+
+  device = Device()
+
+  client.on_connect = device.on_connect
+  client.on_publish = device.on_publish
+  client.on_disconnect = device.on_disconnect
+  client.on_subscribe = device.on_subscribe
+  client.on_message = device.on_message
+
+  client.connect(args.mqtt_bridge_hostname, args.mqtt_bridge_port)
+
+  client.loop_start()
+
+  # This is the topic that the device will publish telemetry events (temperature
+  # data) to.
+  mqtt_telemetry_topic = '/devices/{}/events'.format(args.device_id)
+
+  # This is the topic that the device will receive configuration updates on.
+  mqtt_config_topic = '/devices/{}/config'.format(args.device_id)
+
```

Comment:
It may make sense to acknowledge the messages using the `devices/{}/state` MQTT topic